### ISSUE TITLE
[css-fonts] Update font technology "incremental" into "incremental-patch", "incremental-range", "incremental-auto"

### ISSUE THREAD

**Issue by garretrieger** [w3c/csswg-drafts#7665]
Relevant spec section: https://www.w3.org/TR/css-fonts-5/#font-face-src-parsing

We just updated the incremental font transfer specification to use three font tech keywords ("incremental-patch", "incremental-range", and "incremental-auto") instead of the current single "incremental" keyword (see: https://github.com/w3c/IFT/pull/113, https://w3c.github.io/IFT/Overview.html#opt-in). This change gives the @font-face rule the ability to specify which type of incremental transfer the server hosting the font supports (or auto to instruct the client to figure it out via negotiation).

Could the CSS Fonts spec be updated to match?

**Comment by faceless2** [comment]
We've had support for range `incremental` for some time, as by-product of the use of the same loading approach for PDF. So we're using it server-side, and I appreciate that the main purpose of incremental loading is for mobile or constrained-bandwidth use. 

Based on that I am very, very weakly against this, but mostly I just wanted to add some comments.

Firstly, `range` incremental as I understand it needs no particular support from the font - although the spec recommends that the font is reordered to make Range more efficient, all that's required is for the server to support the `Range` header and the client to support the incremental loading method. The font needs no particular structure internally, other than being OTF or TTF - WOFF, and in particular WOFF2 aren't so useful, as you need the whole table (WOFF) or file (WOFF2) to decompress it.

I'm not familiar with the patch method and we don't support it, but I see it's also done with HTTP headers and looks to be less dependent on the internal font layout.

After a bit of experimenting with `Range` incremental loading, we've started using it for very large fonts where we know in advance that only a few glyphs are going to be used - a page with only a few Chinese glyphs that references `Noto Sans CJK` as the font to display them is pretty much our canonical use case.

In other words we're using `incremental` as a hint to the client to tell it "you'll need this font on the page, but not much of it and it's big".

Given that, from our admittedly aypical perspective I don't see why the method of incremental loading would ever need to be controlled by the author as it's purpose is essentially as a hint as to how the font will be used rather than any specific technical aspect of the font. My question is - under what circumstances would `incremental-patch` be specified and `incremental-range` not wanted?

At the very least if this is approved, could I'd suggest `incremental` instead of `incremental-auto`.

**Comment by svgeesus** [comment]
> I'm not familiar with the patch method and we don't support it, but I see it's also done with HTTP headers and looks to be less dependent on the internal font layout.

The patch-subset method requires an intelligent HTTP server which can on-the-fly create font subsets and then compute a binary patch to upgrade one subset to a (larger) subset. 

And yes, most efficient use of the range-request method needs a re-arranged font as described in the spec while the patch-subset method is unaffected by how the font is laid out.

> My question is - under what circumstances would incremental-patch be specified and incremental-range not wanted?

Patch-subset is [**vastly** better](https://www.w3.org/TR/PFE-evaluation/#analysis-cjk) than range-request, on 3G and faster networks, in terms of bytes transferred, requests made, and overall network cost.

![network cost, CJK](https://www.w3.org/TR/PFE-evaluation/07-08-2020/comparison.cjk.sampled_5000.cost.png)

Stating up-front in the CSS that the server has this capability avoids an initial pre-flight check to ask if the server supports it, so is worth having in the stylesheet. 

It also allows the range-request version to be served from a different server (any HTTP server) while directing patch-subset request to a specialized font CDN.

**Comment by faceless2** [comment]
So this is purely to avoid sending the `Patch` headers in the initial request for the font? Given the obvious benefits, and your very good example of intermediate CDN networks or other proxies that may add support for this, it feels a _little_ odd to make it an author decision to rule out patch-subset at all. But I have plenty of other hills to die on so I'm happy to take your word for it.

**Comment by astearns** [comment]
Are there any other components of a src descriptor that make assertions about the font server environment? I thought they were mainly checks against the UA environment (“Does this browser support incremental fonts? If not, use this other src”).

So these values would be both checking for browser support and also determining how the browser would start talking to the font server. What happens if you have a src that says `incremental-patch` but the server actually does not support it?

**Comment by garretrieger** [comment]
> So this is purely to avoid sending the `Patch` headers in the initial request for the font? Given the obvious benefits, and your very good example of intermediate CDN networks or other proxies that may add support for this, it feels a _little_ odd to make it an author decision to rule out patch-subset at all. But I have plenty of other hills to die on so I'm happy to take your word for it.

When incremental-auto is used the first request has to be a GET request with the request message added as a custom header (in case the server can use them). An unfortunate side affect is that this prevents the request from being considered a simple request for CORS and thus a CORS preflight request may need to get made. So explicitly specifying incremental-patch allows the first request to be made over POST instead and avoid triggering the preflight request. This is a pretty significant performance improvement as it avoids a full round trip. Another benefit of using POST is more compact encoding of the request message (no base64 is needed). For a more detailed discussion on how we arrived at the decision to split the keyword see this writeup: https://docs.google.com/document/d/1IcPg4o5k4OOmfLCfa1z2wQl6MD-kPBrahtcb4r-97ZU/edit?resourcekey=0-9WtXKj_y4uihiB7K740xMQ


> So these values would be both checking for browser support and also determining how the browser would start talking to the font server. What happens if you have a src that says `incremental-patch` but the server actually does not support it?

This is covered by the incxfer spec: https://w3c.github.io/IFT/Overview.html#fallback. If the server responds with something that doesn't look like a patch response then the client will just perform a normal non-incremental load on that URL. An alternative behaviour we could consider specifying instead would be to have the client skip that url and try the next URL in the @font-face (if present).

**Comment by cdoublev** [comment]
Thanks for the [change](https://github.com/w3c/csswg-drafts/commit/a951f1f2e69e6945c6d44e7eca5809a3e15a8485) Chris. Do you think this can be applied to [`<font-tech>` in CSS Conditional 5](https://drafts.csswg.org/css-conditional-5/#typedef-font-tech) as well?

`w3c/reffy` also extracts the definition from [CSS Fonts 4](https://drafts.csswg.org/css-fonts-4/#font-tech-values) and [5](https://drafts.csswg.org/css-fonts-5/#font-tech-values). Ideally, there should be only one definition. Its users have to know which spec has authority, which happens to be non-trivial in this case.

CSS Conditional 5 also have its own definition of `<font-format>` and a note saying that `<font-format>` and `<font-tech>` should be imported from CSS Fonts. But in `@supports` it actually excludes `<string>` (cf. https://github.com/w3c/csswg-drafts/issues/8110#issuecomment-1330516749), therefore some may think that `incremental` matches `<font-tech>` but not `incremental-*` in `@supports`, and vice-versa in `src`.
